nats.go

```diff
@@ -2601,7 +2629,41 @@ func (nc *Conn) kickFlusher() {
 // argument is left untouched and needs to be correctly interpreted on
 // the receiver.
 func (nc *Conn) Publish(subj string, data []byte) error {
-	return nc.publish(subj, _EMPTY_, data)
+	return nc.publish(subj, _EMPTY_, nil, data)
+}
+
+// Used to create a new message for publishing that will use headers.
+func NewMsg(subject string) *Msg {
+	return &Msg{
+		Subject: subject,
+		Header:  make(http.Header),
+	}
+}
+
+const (
+	hdrLine   = "NATS/1.0\r\n"
+	crlf      = "\r\n"
+	hdrPreEnd = len(hdrLine) - len(crlf)
+)
+
+// decodeHeadersMsg will decode and headers.
+func decodeHeadersMsg(data []byte) (http.Header, error) {
+	// FIXME(dlc) - brittle and slow probably.
+	br := bufio.NewReader(bytes.NewBuffer(data))
+	hl, err := br.ReadString('\n')
+	if err != nil {
+		return nil, err
+	}
+	// FIXME(dlc) - pull status and description?
+	if !strings.HasPrefix(hl, hdrLine[:hdrPreEnd]) {
+		return nil, ErrBadHeaderMsg
+	}
+	tp := textproto.NewReader(br)
+	mh, err := tp.ReadMIMEHeader()
+	if err != nil {
+		return nil, ErrBadHeaderMsg
+	}
```

Comment:
Any chance we can get at how many bytes the reader consumed and check to see if that matches the header/msg boundary

Reply:
Not sure what the ReadMIMEHeader() does but like I said may unpack or at least at a check for the trailing `\r\n\r\n`

Reply:
I cleaned it up a bit.